sdk/src/lib.rs

```diff
@@ -743,6 +740,39 @@ impl WorkflowFunction {
             }),
         }
     }
+
+    /// Build a workflow function from a closure or function pointer which accepts a [WfContext] and takes one argument
+    pub fn from<A, F, Fut, R, O>(f: F) -> Self
```

Comment:
So ideally both this and the activity version are implemented generically (as actual `From` impls) for anything that matches these bounds -- allowing us to avoid `XXXFunction::from` calls at definition sites, since the registration functions already accept `impl Into<XXXFunction>`. I think this is a really nice piece of ergonomics to maintain.

Reply:
Hi, Sorry, did not understand this comment completely.

Reply:
The idea would be to do something like this:

```
impl< ... generics ... > From<F> where: bounds for ActivityFunction {
   ...
}
```
For both activity and wf functions

So there's no need to call `ActivityFunction::from` at the same place you're defining or registering the function - everything just slots in.

Reply:
Haa,

I tried to do that, but the From implementation cannot be implemented for function with generics, it fails with '[unconstrained](https://stackoverflow.com/questions/68639821/unconstrained-type-parameter-for-impl-fromstruct-for-t-when-impl-struct) bounds'

So the only From implementation is the one [existing](https://github.com/temporalio/sdk-core/blob/master/sdk/src/lib.rs#L710). Activity/Workflow Function that takes only Context.

For functions that take an argument, We need the `ActivityFunction::from/WorkflowFunction::new` [wrapper](https://github.com/temporalio/sdk-core/blob/master/core/src/core_tests/child_workflows.rs#L103).

But in the production SDK, we will actually not use these functions(I think). There should be higher-level 'execute_activity' or 'execute_childworkflow' functions that exist in other SDKs but that check the types of the arguments like [this](https://github.com/h7kanna/sdk-core/blob/workflow-definitions/tests/integ_tests/workflow_tests/examples.rs#L296) 

Essentially, the PR is to allow any function that can be converted into 

```
dyn Fn(WfContext) -> BoxFuture<'static, Result<WfExitValue<Payload>, anyhow::Error>>
    + Send
    + Sync
    + 'static
```
or
```
    dyn Fn(ActContext) -> BoxFuture<'static, Result<ActExitValue<Payload>, anyhow::Error>>
        + Send
        + Sync,
```

Reply:
So this https://github.com/temporalio/sdk-core/blob/master/sdk/src/lib.rs#L840 implements an arg-accepting automatic implementation for activities - and, yeah, clearly we can't make that work for both 0 and 1 input-arity functions without some extra boilerplate or macros, etc. That said, there's nothing preventing implementing `IntoActivityFunc` for 0 and 1 input-arity functions and having this all work properly. Same could be done with workflows.

I think that'd be the best place to leave things for now, no need to worry about higher-arity functions at the moment.

You can see a preview of what I'm thinking of for later here: https://github.com/temporalio/sdk-core/pull/550 which will lean on procmacros for making this all seamless.

Reply:
I will look up the proposal and adapt this accordingly.